0004/README.md

```diff
@@ -0,0 +1,524 @@
+update must not invalidate previous 'good' versions of that package. The
+solution is to qualify expectations for a package with a version range.
+
+-  With alternative 1,  different version number ranges correspond to
+    different expected provenance values .
+-  With alternative 2, different version number ranges correspond to
+    different past 'good' provenance instances.
+
+#### Comparison of alternatives
+
+-  Both alternatives could work whether managed locally and centrally. 
+-  Both are compatible with the "trust on first use" approach described
+    below, because both generate provenance expectations, based on:
+    -  A configurable set of provenance properties that are not
+        expected to change. This could be an organization-wide or even
+        ecosystem-wide default setting.
+    -  The content of a 'good' provenance instance. 
+
+I am inclined to prefer alternative 1, because it is more explicit about what
+provenance will or will not be compliant. Alternative 2 may be preferable if
+only information about the last package version needs to be remembered.
+
+## Bootstrapping policy support
+
+### Generating provenance expectations
+
+To overcome a bootstrapping problem, we assume that compromises will be rare,
+and propose to generate package-specific SLSA provenance expectations on first
+use. Each such 'trust on first use' (TOFU) policy contains expectations based on
+observed provenance, and can be used as a baseline to detect unexpected changes,
+for example a too low SLSA level, missing provenance, or an unexpected change in
+a 'sensitive' provenance property. A [proposed TOFU
+policy](#a-proposed-tofu-policy-for-a-single-npm-package) is presented below.
+
+### From local to centralized management
+
+We expect that provenance expectations will initially be managed locally where
+they are used: at the package developer before publishing to a registry, at the
+package registry, or at the package user. As experience grows and confidence
+increases, policies may be migrated to a central registry and managed there.
+
+Future work is to implement centralized policy management.
+
+### From alerting-only to enforcing
+
+We expect that initially the policy evaluator will alert for an unexpected
+change, and not block it.
+
+-  An auditing policy can be evaluated post-upload and post-installation.
+    This minimizes interference with existing npm tooling. 
+-  An enforcing policy needs to be more tightly integrated with npm tools,
+    so that it can block a non-compliant package.
+
+Future work is to integrate the policy evaluator with npm tooling.
+
+### A proposed TOFU policy for a single npm package
+
+This is for a policy that lists explicit expectations for provenance values,
+instead of expectations for what provenance fields are invariant (regardless of
+their value).
+
+Inputs: 
+
+-  Package name and version.
+-  Optional: verified package provenance.
+-  Preferences. Examples include, but are not limited to:
```

Comment:
Specifically, I guess this would include things like
- repository url
- builder id (without version suffix?)
- repository id
- owner id
- workflow ref, job workflow ref (without the branch name?)

Seems like we wouldn't be able to match on tag or branch name as this might cause a lot of false positives when people use release branches. Event name might also be a problem if you legitimately use two events to publish.

Reply:
Yes, there needs to be a way to constrain what claims provenance can make, with some granularity that is still TBD. A possible action could be to downgrade a package to a low SLSA level or to reject it.